### PR TITLE
GEODE-9792: synchronize multi-user authentication on different threads

### PR DESCRIPTION
**Pull request by jinmeiliao**
Avoid making multiple authentication requests in multi-user case.

Also remove some unnecessary code and small refactoring.

### CODE REVIEW

File: geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java [hunk 1]
```diff
@@ -252,4 +283,50 @@ public void jsonFormatterOnTheClientWithMultiUser() throws Exception {
...
+        .createCache();
+    Properties properties = new Properties();
+    properties.setProperty(UserPasswordAuthInit.USER_NAME, "data");
+    properties.setProperty(UserPasswordAuthInit.PASSWORD, "wrongPassword");
+    RegionService regionService = cache.createAuthenticatedView(properties);
+
+    cache.createClientRegionFactory(ClientRegionShortcut.PROXY).create("region");
+    Region region = regionService.getRegion(SEPARATOR + "region");
+    assertThatThrownBy(() -> region.registerInterestForAllKeys())
+        .isInstanceOf(UnsupportedOperationException.class);
```
Comment (DonalEvans) [inline geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java:325]: This assertion seems unrelated to the code being tested in this class, since `registerInterest()` will always throw an `UnsupportedOperationException` when called on a `ProxyRegion`, regardless of the authentication being used. Can it just be removed?

File: geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java [hunk 2]
```diff
@@ -252,4 +283,50 @@ public void jsonFormatterOnTheClientWithMultiUser() throws Exception {
...
+        .withProperty(SECURITY_CLIENT_AUTH_INIT, CountableUserPasswordAuthInit.class.getName())
+        .withMultiUser(true)
+        .createCache();
+    Properties properties = new Properties();
+    properties.setProperty(UserPasswordAuthInit.USER_NAME, "data");
+    properties.setProperty(UserPasswordAuthInit.PASSWORD, "wrongPassword");
+    RegionService regionService = cache.createAuthenticatedView(properties);
+
+    cache.createClientRegionFactory(ClientRegionShortcut.PROXY).create("region");
+    Region region = regionService.getRegion(SEPARATOR + "region");
```
Comment (DonalEvans) [inline geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java:323]: Warning here can be fixed by using `Region<Object, Object>`

File: geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java [hunk 3]
```diff
@@ -226,12 +231,38 @@ public void jsonFormatterOnTheClientWithSingleUser() throws Exception {
     region.put("key", value);
 
     // make sure the client only needs to authenticate once
-    assertThat(TrackableUserPasswordAuthInit.timeInitialized.get()).isEqualTo(1);
+    assertThat(CountableUserPasswordAuthInit.count.get()).isEqualTo(1);
   }
 
+  @Rule
+  public ExecutorServiceRule executor = new ExecutorServiceRule();
```
Comment (DonalEvans) [inline geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java:238]: Can this rule be put at the top of the class with the other rules please?

File: geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java [hunk 4]
```diff
@@ -226,12 +231,38 @@ public void jsonFormatterOnTheClientWithSingleUser() throws Exception {
...
+        .withProperty(SECURITY_CLIENT_AUTH_INIT, CountableUserPasswordAuthInit.class.getName())
+        .withMultiUser(true)
+        .createCache();
+    Properties properties = new Properties();
+    properties.setProperty(UserPasswordAuthInit.USER_NAME, "data");
+    properties.setProperty(UserPasswordAuthInit.PASSWORD, "data");
+    RegionService regionService = cache.createAuthenticatedView(properties);
+
+    cache.createClientRegionFactory(ClientRegionShortcut.PROXY).create("region");
+    Region region = regionService.getRegion(SEPARATOR + "region");
```
Comment (DonalEvans) [inline geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java:252]: Warning here can be fixed by using `Region<Object, Object>`

File: geode-core/src/test/java/org/apache/geode/cache/client/internal/OpExecutorImplUnitTest.java [hunk 5]
```diff
@@ -94,9 +95,57 @@ public void authenticateIfRequired_setId_singleUser_hasNoId() {
...
+    when(connection.execute(any())).thenReturn(123L);
+    when(connectionManager.borrowConnection(5)).thenReturn(connection);
+    OpExecutorImpl spy = spy(executor);
+
+    spy.execute(op, 1);
+    verify(spy).authenticateIfMultiUser(any(), any());
   }
 
+  @Test
+  public void authenticateIfMultiUser_calls_authenticateMultiUser() {
```
Comment (DonalEvans) [inline geode-core/src/test/java/org/apache/geode/cache/client/internal/OpExecutorImplUnitTest.java:113]: This test seems to be testing a lot of different behaviour. It might be best to split it out into multiple tests that are each testing one thing, so that if there's a failure, it's more obvious at a glance what the problem is.

**Review by DonalEvans — CHANGES_REQUESTED** [review]

File: geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java [hunk 6]
```diff
@@ -252,4 +283,50 @@ public void jsonFormatterOnTheClientWithMultiUser() throws Exception {
...
+      region.put("key1", "value1");
+      region.put("key2", "value2");
+      txManager.commit();
+    } catch (CommitConflictException conflict) {
+      // ... do necessary work for a transaction that failed on commit
+    } finally {
+      if (txManager.exists()) {
+        txManager.rollback();
+      }
+    }
```
[the remark this answers is not in the document]
Comment (jinmeiliao) [reply geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java:307]: the test was added for code understanding. Not needed.

File: geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java [hunk 7]
```diff
@@ -252,4 +283,50 @@ public void jsonFormatterOnTheClientWithMultiUser() throws Exception {
...
+  }
+
+  @Test
+  public void multiUserWithClientSubscription() throws Exception {
```
[the remark this answers is not in the document]
Comment (jinmeiliao) [reply geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java:311]: the test was added for code understanding. Not needed, but kept since it provide some coverage.

File: geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java (hunk 1 above)
Comment (jinmeiliao) [reply geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java:325]: the test was added for code understanding. Not needed.

File: geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java (hunk 2 above)
Comment (jinmeiliao) [reply geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java:323]: removed

File: geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java (hunk 3 above)
Comment (jinmeiliao) [reply geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java:238]: done

File: geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java (hunk 4 above)
Comment (jinmeiliao) [reply geode-cq/src/distributedTest/java/org/apache/geode/security/MultiUserAPIDUnitTest.java:252]: removed

File: geode-core/src/test/java/org/apache/geode/cache/client/internal/OpExecutorImplUnitTest.java (hunk 5 above)
Comment (jinmeiliao) [reply geode-core/src/test/java/org/apache/geode/cache/client/internal/OpExecutorImplUnitTest.java:113]: I was always torn between both ways: split into multiple tests or put them in one and have less code. If test fails, which line it fails definitely tells you what exactly the problem is.  This way, it seems very concise and provide a stream line of the stepping further down the code.

**Review by kirklund — APPROVED** [review]
